[SYCL][ESIMD] Add tests on simd copy and move assignment operators

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add tests on simd copy and move assignment operators

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment_copy_core.cpp:
```diff
@@ -0,0 +1,118 @@
+//
+// Test for simd copy assignment operator.
+// The test creates source simd instance with reference data and invokes copy
+// assignment operator. It is expected that simd object that called copy
+// assignment operator will stor a bitwise same data as in the source simd
+// object.
```
```suggestion
// assignment operator from source to destination simd instance. It is expected
// for destination simd instance to store a bitwise same data as the reference one
```

vasilytric replied:
I've done it in this commit: c08d8bcd8cfb2adfcd4af2321500e8672a2b471c

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment_copy_core.cpp:
```diff
@@ -0,0 +1,118 @@
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional;
+
+// Descriptor class for the case of calling constructor in initializer context.
```
There is no initializer context below. Anyway, looks like we don't have a `context` for assignment operators.
As a suggestion, what if we have:

- cpp file with `copy_assignment`
- cpp file with `move_assignment`
- header file with generic `run_test` for move and copy assignment

Would it be reasonable? Given that:

- it would be great to check that `was_move_destination() == false` for copy assignment
- we might want someday to check copy assignment doesn't invalidate source simd, but we cannot allow such check for move assignment

vasilytric replied:
I've done it in this commit: c08d8bcd8cfb2adfcd4af2321500e8672a2b471c

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment_move_core.cpp:
```diff
@@ -0,0 +1,132 @@
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         allocator);
+
+    shared_vector<int> was_moved(1, shared_allocator<int>(queue));
```
Maybe we need to define a separate RAII wrapper for non-containers, what do you think? To avoid using std::vector in case we need a single USM variable

vasilytric replied:
we can use a smart pointer that can delete used heap memory, but I suggest do it if it will be useful in many cases and now don't change it

vasilytric replied:
@yuriykoch I've done it in this commit: 8180aea8bfbdf9fcfd20406eec5a97c0b2444f25
Was added `shared_elements` that lets interact with USM by smart pointer

vasilytric committed:
[SYCL][ESIMD] Update file structure

- Use header with werifecation function it lets avoid code copy-paste
- Update comments

vasilytric committed:
[SYCL][ESIMD] Add shared element that lets interact with USM

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -14,15 +14,15 @@
 #include "logger.hpp"
 #include "type_coverage.hpp"
 #include "type_traits.hpp"
 #include "value.hpp"
 
+#include <sycl/ext/intel/experimental/esimd.hpp>
+#include <sycl/sycl.hpp>
+
 #include <vector>
```
Why such move of includes? By doing so we might have something from our headers affecting the esimd api itself.

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R17-R26)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment.hpp:
```diff
@@ -0,0 +1,97 @@
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         allocator);
+
+    shared_elements<DataT> was_moved(queue);
```
```suggestion
    shared_element<DataT> was_moved(queue);
```
Regarding naming. It's a single shared variable of a type DataT, so no need in plural form here.

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-c5eea7de2a62b10b21518cace82c8c9c376b703b2d585f6f25b9bfd4ef1ee8a4R60)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment.hpp:
```diff
@@ -0,0 +1,97 @@
+#include "../shared_element.hpp"
+#include "common.hpp"
+
+namespace esimd_test::api::functional::ctors {
```
Wait a minute. Why do we have a tests for `operator` in `constructors` namespace and folder?

vasilytric replied:
I've done it in this commit: 4be7ac99eca7778991263036b9340b4580e5e378

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment.hpp:
```diff
@@ -0,0 +1,97 @@
+
+#pragma once
+
+// The test proxy is used to verify the move constructor was called actually.
```
```suggestion
// The test proxy is used to verify the move assignment was called actually.
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-c5eea7de2a62b10b21518cace82c8c9c376b703b2d585f6f25b9bfd4ef1ee8a4R17)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment.hpp:
```diff
@@ -0,0 +1,97 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides functios for tests on simd assignment operators.
```
```suggestion
/// This file provides functions for tests on simd assignment operators.
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-c5eea7de2a62b10b21518cace82c8c9c376b703b2d585f6f25b9bfd4ef1ee8a4R11)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment.hpp:
```diff
@@ -0,0 +1,97 @@
+
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename DataT, int NumElems, typename TestCaseT,
+          typename ShouldMoveCtorBeCalled>
```
```suggestion
template <typename DataT, int NumElems, typename TestCaseT>
```
Isn't a requirement to call or not to call `move assignment operator` a part of the test case, as result - a part of `testCaseT` interface?
Please, verify you don't mention constructors instead of operators anywhere.

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-c5eea7de2a62b10b21518cace82c8c9c376b703b2d585f6f25b9bfd4ef1ee8a4L27-L30)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment.hpp:
```diff
@@ -0,0 +1,97 @@
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         allocator);
+
+    shared_elements<DataT> was_moved(queue);
```
Regarding logic.

- Please, consider providing value to initialize the `was_moved` flag explicitly
- Why is it `DataT`, not the `int` or `bool`?

vasilytric replied:
I've done it in this commit: cdfde0ae290eadbcc580fb2aa548dd5cad143c8f

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment.hpp:
```diff
@@ -0,0 +1,97 @@
+      }
+    }
+
+    if (was_moved[0] != ShouldMoveCtorBeCalled::value) {
```
```suggestion
    if (was_moved.value() != ShouldMoveCtorBeCalled::value) {
```
As a suggestion regarding `was_moved` usage

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-c5eea7de2a62b10b21518cace82c8c9c376b703b2d585f6f25b9bfd4ef1ee8a4R86)

yuriykoch commented on SYCL/ESIMD/api/functional/shared_elements.hpp:
```diff
@@ -0,0 +1,43 @@
+        sycl::malloc_shared<T>(num_elements, device, context), deleter);
+  }
+
+  T *data() { return m_allocated_data.get(); }
+
```
```suggestion
  T *data() { return m_allocated_data.get(); }

  const T *data() const { return m_allocated_data.get(); }
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-9d2d5320f066e81e77b17efb5f2888df4d6d5e653f591eb2e9d39f34c267172cR40)

yuriykoch commented on SYCL/ESIMD/api/functional/shared_elements.hpp:
```diff
@@ -0,0 +1,43 @@
+  std::unique_ptr<T, std::function<void(T *)>> m_allocated_data;
+
+public:
+  shared_elements(sycl::queue &queue, size_t num_elements = 1) {
```
OK, I see now, it's just another way to define `shared_vector` of a fixed size. I was talking about scalar variables to get rid of `operator[]` though..

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-9d2d5320f066e81e77b17efb5f2888df4d6d5e653f591eb2e9d39f34c267172cL40-R42)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_assignment_copy_core.cpp:
```diff
@@ -0,0 +1,61 @@
+    simd<DataT, NumElems> simd_obj;
+    simd_obj = source_simd;
+    simd_obj.copy_to(out);
+    return simd_obj.get_test_proxy().was_move_destination();
```
```suggestion
    return simd_obj.get_test_proxy().was_move_destination() == false;
```
No need to set `false` for `copy` and `true` for `move` on the upper level of abstraction, right? Don't forget to rename `was_moved` in `run_test` appropriately

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/762/commits/afbf175e94a84d6a621ba620d2c0119abd0ec6aa#diff-4f695945bc5522c650a312a3ef5b0a9fbacce7dc658eb84d8ebad6754bff46a6R40)

vasilytric committed:
[SYCL][ESIMD] Update shared_elem and small changes

- Update comemnts
- Now shared_elem used only with one elem

vasilytric committed:
[SYCL][ESIMD] Move tests on operators into separate folder

Also there was different changes according to move TestDescription into
ctors/common.hpp

vasilytric committed:
[SYCL][ESIMD] Use bool for shared_element

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -46,6 +46,38 @@ template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
+    std::string log_msg("Failed for simd<");
+
+    log_msg += m_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg += ", with context: " + ContextT::get_description();
```
Such output not required for copy and move assignment operators. Maybe it would be better to have a separate `TestDescription`-s for constructors and operators for now? The comment is minor, still it might result in reverting multiple small changes in future

vasilytric replied:
I don't see differences between output for this two cases, and can't imagine they. I suggest to do it when we can see these differences

yuriykoch replied:
Regarding the semantic difference, imagine the following:
* `ctor_copy` fails `with context: variable declaration`
* `ctor_copy` fails `with context: rvalue in an expression`
* `ctor_move` fails `with context: variable declaration`
* `ctor_move` fails `with context: rvalue in an expression`
* `operator_assignment_copy` fails `with context: copy assignment` (every time)
* `operator_assignment_move` fails `with context: move assignment` (every time)

Looks like for assignment operators there is no strict need in `with context:` string

Once again, its a minor comment, so feel free to choose any way you see appropriate

vasilytric replied:
@yuriykoch after some thinking I agreed with you, I've done it in this commit: 85f66a6afee16d8461b96d0739f36ca46c24aefe

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_assignment.hpp:
```diff
@@ -0,0 +1,96 @@
+#include "../common.hpp"
+#include "../shared_element.hpp"
+
+namespace esimd_test::api::functional::ctors {
```
```suggestion
namespace esimd_test::api::functional::operators {
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-edf690369570284c343904e24bdf038b8c4bb8e6b620e467934162c5cdd1e05eR23)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_assignment.hpp:
```diff
@@ -0,0 +1,96 @@
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         allocator);
+
+    shared_element<bool> was_moved(queue);
```
Looks like the naming should be changed. It stores some `success` now, right? So for both copy and move assignment `true` is expected.
Once again, it's better to have some initial value provided explicitly

```suggestion
    shared_element<bool> is_correct_operator(queue, false);
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-edf690369570284c343904e24bdf038b8c4bb8e6b620e467934162c5cdd1e05eR60)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_assignment.hpp:
```diff
@@ -0,0 +1,96 @@
+      }
+    }
+
+    if (!was_moved.value()) {
+      passed = false;
+      log::note("Test failed due to move assignment operator hasn't called for "
+                "simd<" +
+                data_type + ", " + std::to_string(NumElems) + ">.");
+    }
```
Let's have something that may fit for both operators. Output would be strange for copy assignment.

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-edf690369570284c343904e24bdf038b8c4bb8e6b620e467934162c5cdd1e05eR87-R89)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_assignment.hpp:
```diff
@@ -0,0 +1,96 @@
+      cgh.single_task<Kernel<DataT, NumElems, TestCaseT>>(
+          [=]() SYCL_ESIMD_KERNEL {
+            *was_moved_storage =
+                TestCaseT::template call_simd_ctor<DataT, NumElems>(ref, out);
```
```suggestion
                TestCaseT::template run<DataT, NumElems>(ref, out);
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-edf690369570284c343904e24bdf038b8c4bb8e6b620e467934162c5cdd1e05eR70)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_assignment_copy_core.cpp:
```diff
@@ -0,0 +1,58 @@
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional;
+
+// Descriptor class for the case of calling constructor in initializer context.
```
Update please as discussed [earlier](https://github.com/intel/llvm-test-suite/pull/762#discussion_r791622390)

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-5ae6688531ef20b0d69bf7dd54176b9516394d4266983c2876a387324a402c15R29)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_assignment_move_core.cpp:
```diff
@@ -0,0 +1,59 @@
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional;
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct move_assignment {
```
Update please

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d)

yuriykoch commented on SYCL/ESIMD/api/functional/shared_element.hpp:
```diff
@@ -0,0 +1,45 @@
+
+    auto deleter = [=](T *ptr) { sycl::free(ptr, context); };
+
+    m_allocated_data = std::unique_ptr<T, decltype(deleter)>(
+        sycl::malloc_shared<T>(1, device, context), deleter);
+  }
```
```suggestion
    m_allocated_data = std::unique_ptr<T, decltype(deleter)>(
        sycl::malloc_shared<T>(1, device, context), deleter);
    assert(m_allocated_data && "USM memory allocation failed");
    *m_allocated_data = initial_value;
  }
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-9d2d5320f066e81e77b17efb5f2888df4d6d5e653f591eb2e9d39f34c267172cR37-R38)

yuriykoch commented on SYCL/ESIMD/api/functional/shared_element.hpp:
```diff
@@ -0,0 +1,45 @@
+  std::unique_ptr<T, std::function<void(T *)>> m_allocated_data;
+
+public:
+  shared_element(sycl::queue &queue) {
```
```suggestion
  shared_element(sycl::queue &queue, T initial_value) {
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-9d2d5320f066e81e77b17efb5f2888df4d6d5e653f591eb2e9d39f34c267172cR28)

yuriykoch commented on SYCL/ESIMD/api/functional/shared_element.hpp:
```diff
@@ -0,0 +1,45 @@
+
+  const T *data() const { return m_allocated_data.get(); }
+
+  T value() { return *m_allocated_data.get(); }
```
```suggestion
  T value() const { return *m_allocated_data.get(); }
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-9d2d5320f066e81e77b17efb5f2888df4d6d5e653f591eb2e9d39f34c267172cR47)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_assignment.hpp:
```diff
@@ -0,0 +1,96 @@
+
+// The test proxy is used to verify the move assignment was called actually.
+#define __ESIMD_ENABLE_TEST_PROXY
+
+#include "../common.hpp"
+#include "../shared_element.hpp"
+
```
Do we really need to distinguish the place to include `shared_element` and `shared_vector`? That's the same level of abstraction, so looks like we might just include shared_element into the same header as the shared_vector

vasilytric replied:
`shared_element` not needed in all tests, so this inclusion will be useless in such cases.
Therefore I suggest do not include in in SYCL\ESIMD\api\functional\ctors\common.hpp

vasilytric committed:
[SYCL][ESIMD] Small updates

- Rename ctors into operators
- Rename variables
- Add assert tu verify that USM was allocated correctly
- Add possible to specify init value to shared_element

vasilytric commented:
CI on Windows platform failed due to issue, that is not related to this PR (failed Plugin/level_zero_device_scope_events.cpp tests) this test disabled, but fix not merged yet.
@kbobrovs Would you mind to take a look?

vasilytric committed:
Merge branch 'intel' into copy_move_assgnment

vasilytric committed:
[SYCL][ESIMD] Use separate common files for tests on ctors and operators

Revert ctors/common.hpp that has deleted in this PR

vasilytric committed:
Merge branch 'intel' into copy_move_assgnment

vasilytric commented:
@intel/dpcpp-tools-reviewers @intel/llvm-reviewers-runtime @steffenlarsen 
You all were added by accident. Please, ignore all notifications from this PR

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

vasilytric committed:
[SYCL][ESIMD] Use for_all_combinations instead of for_all_types_and_dims

kbobrovs commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -46,6 +46,9 @@ template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
 
 } // namespace details
 
+// Dummy kernel for submitting some code into device side.
```
```suggestion
// Class used as a kernel ID.
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/f6c0e9809f2bbdef3c23496b8f2dd8e8c5e07a88#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0L49-R49)

kbobrovs commented on SYCL/ESIMD/api/functional/operators/common.hpp:
```diff
@@ -0,0 +1,51 @@
+
+namespace esimd_test::api::functional::operators {
+
+// Dummy kernel for submitting some code into device side.
```
ditto

vasilytric replied:
Unused kernels has been removed in this commit: e8b81483455f0a82e3d719185b184e3c2dd1715c

kbobrovs commented on SYCL/ESIMD/api/functional/operators/operator_assignment_move_core.cpp:
```diff
@@ -0,0 +1,61 @@
+using namespace esimd_test::api::functional;
+
+// Descriptor class for the case of calling move assignment operator.
+struct move_assignment {
```
Nit: I would have move and copy ctors in the same source as they share a lot

vasilytric replied:
I've applied this comment in this commit: a80c2e39b68ed61f908a5454b77590dbc907f510

kbobrovs commented on SYCL/ESIMD/api/functional/shared_element.hpp:
```diff
@@ -0,0 +1,50 @@
+
+namespace esimd_test::api::functional {
+
+template <typename T> class shared_element {
```
please add some docs how this struct achieves the stated goals

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/f6c0e9809f2bbdef3c23496b8f2dd8e8c5e07a88#diff-9d2d5320f066e81e77b17efb5f2888df4d6d5e653f591eb2e9d39f34c267172cR25-R27)

vasilytric committed:
[SYCL][ESIMD] Unite tests on copy and move assignment operators into one file

vasilytric committed:
[SYCL][ESIMD] Update comments

vasilytric committed:
[SYCL][ESIMD] Remove unused kernel definitions

vasilytric committed:
Merge branch 'intel' into copy_move_assgnment

vasilytric committed:
[SYCL][ESIMD] Replace ctors::Kernel with Kernel
